packages/grafana-data/src/types/datasource.ts

```diff
@@ -310,12 +320,13 @@ export interface MetadataInspectorProps<
 export interface QueryEditorProps<
   DSType extends DataSourceApi<TQuery, TOptions>,
   TQuery extends DataQuery = DataQuery,
-  TOptions extends DataSourceJsonData = DataSourceJsonData
+  TOptions extends DataSourceJsonData = DataSourceJsonData,
+  TVQuery extends DataQuery = TQuery
 > {
   datasource: DSType;
-  query: TQuery;
+  query: TVQuery;
   onRunQuery: () => void;
-  onChange: (value: TQuery) => void;
+  onChange: (value: TVQuery) => void;
```

Comment:
This looks a bit weird to me. Do we need both TQuery and TVQuery? Like is there any moment you would specify both? I assume not because in that case TQuery won't be actually used anywhere.

Reply:
You're right, but unfortunately TQuery is inferred by DataSourceAPI so you can't tell the QueryEditor to use a different query than the one declared in DataSourceAPI. We need this for custom variable support where you can have a custom TVQuery. 

But I'm opened for other ways to solve this. I started by creating my own QueryEditorProps for variables, but that would mean they would have to be in sync.